### PR TITLE
Block `SequentialQueue` on reauthentication. Reorganize Network / Reauthentication code.

### PR DESCRIPTION
**Pull request by marcaaron**
### Details

Makes the `SequentialQueue` work better when we are reauthenticating.

**Before:** When a request in the SQ needed reauthentication we would call Authenticate and then send the original request into the non-sequential queue (where it would be processed after the SQ empties). That is bad because requests happen out of order.

```
Request 1 (Fail)
Authenticate
Request 2
Request 3
Request 1
```

**After:** The reauthentication logic has been reorganized and calls to Authenticate will block the queue

```
Request 1 (Fail)
Authenticate
Request 1 
Request 2
Request 3
```

### Fixed Issues
https://github.com/Expensify/Expensify/issues/207534

### Tests (Build staging version of app)
1. Go offline via Chrome Dev Tools Network tab

3. Leave 10 comments on a report
4. Invalidate the authToken via Preferences

5. Go back online
6. Verify by inspecting the Network requests tab in Chrome Dev Tools that each request in the queue is made sequentially and that the first request requiring reauthentication will wait for `Authenticate` to complete before moving onto the next request in the queue.

- [ ] Verify that no errors appear in the JS console

### PR Review Checklist

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

**Comment by marcaaron** [comment]
@tgolen This PR got a little out of hand - but I think in a good way. 

So many things that were bugging me about the Network code feel pretty nice now with the middleware idea applied.

Here's the basic idea...

- `Network.post()` passes requests to the queues
- The queues call `Request.process()` which calls `HttpUtils.xhr()` with middlewares applied
- Each middleware returns a promise and passes the response to the next middleware
- The order is important and we have:

    - Logging - Logs the request details (this cleaned up all circular dependencies with Log)
    - Recheck connection - Not sure how much we need the stuff in this file, but figured I'd leave it for now.
    - Reauthentication - Handles reauthentication
    - Retry - Handles request retry logic

One thing that is still bugging me about this (but will be fixed shortly) is that I've got a `isFromSequentialQueue` argument getting passed to the middleware. This is necessary because I haven't yet forced all network requests that can be persisted directly into the sequential queue. Once we do that we can just look at the request object and see if it's "persistable", but for now, `isFromSequentialQueue` is the only  way to tell if it was called from the sequential queue or not.

File: src/libs/Middleware/Reauthentication.js
Comment (tgolen) [inline src/libs/Middleware/Reauthentication.js:48]: What do you think about moving this to `Authenticate.isAuthenticating()`?

File: src/libs/Middleware/Reauthentication.js
Comment (tgolen) [inline src/libs/Middleware/Reauthentication.js:53]: I think I agree with this solution but I'm not sure what that might imply.

**Review by tgolen — COMMENTED** [review]
Overall I think this looks really clean and organized. I like how the middleware pattern keeps logic very self-contained and purposeful.

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:53]: Yeah, I haven't quite worked it out, but I think it'd be something like...

- We only ever have one call to Authenticate happening at a time and we have a reference to that promise in the middleware
- If a request is made and that promise hasn't resolved yet we wait until it does by hooking into it and using it's response

I will caveat that I also haven't observed the case I'm describing happen, but did write a failing test that seems to imply it could happen like this:

- Request made while online returns 407 which triggers a call to `Authenticate` to get queued, but not run
- We go offline the same time the response is received but before we actually call `Authenticate` (not sure if this is possible)
- We make some requests while offline
- We get "stuck" `isAuthenticating` (again, not sure if it's possible)
- The Sequential Queue runs and the first request gets a 407
- We can't continue because we are "already authenticating"

This is why I had the instinct to throw an error - which would reveal if it's possible or not.

Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:53]: Actually, the situation I am describing can be prevented by just checking to see if we are offline when handling a 407 and throwing an error if we are. That seems like a simpler solution so gonna try that.

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:53]: One idea I am having is to create an `AuthenticationUtils`. 

I've tried a couple different things and didn't land on how to fix the dependency cycle (I need more practice fixing those puzzles 😄 ), but it maybe is reasonable to have `authenticate` and `reauthenticate` in `Authentication` and put the credentials and isAuthenticating stuff in an `AuthenticationUtils`

Comment (tgolen) [reply src/libs/Middleware/Reauthentication.js]: Ah, circular dependencies can be a huge pain like that. I think I'd rather leave them on `NetworkStore` rather than having a new `AuthenticationUtils`.

It's at least better that `Network` doesn't have any specialized knowledge now, but it is a little hard trying to reason about the difference between `Network` and `NetworkStore`.

File: src/libs/Middleware/Reauthentication.js
Comment (tgolen) [reply src/libs/Middleware/Reauthentication.js]: Ah, cool. In the end, if we still really want to log for this case, then I think we can just leave a comment that says it's there for a bit to ensure that nothing ever happens (because if it did happen, we would never know it, and it would lead to bugs that would be very hard to debug) and maybe we can go in later and remove the error.

**Comment by marcaaron** [comment]
Ok ready for review now. 
Gonna add another reviewer here since the changes are pretty extensive (though a lot of this is just reorganizing). 
I am open to splitting this stuff up into several PRs, but would have to think about how to do it.

File: src/libs/API.js
Comment (tgolen) [inline src/libs/API.js:17]:
```suggestion
// Reauthentication - Handles jsonCode 407 which indicates an expired authToken. We need to reauthenticate and get a new authToken with our stored credentials.
```

File: src/libs/API.js
Comment (tgolen) [inline src/libs/API.js:14]:
```suggestion
// Recheck - Sets a timer for a request that will "recheck" if we are connected to the internet if time runs out. Also triggers the connection recheck when we encounter any error.
```

File: src/libs/API.js
Comment (tgolen) [inline src/libs/API.js:15]: Suggestion: rename to `Middleware.RecheckConnection`

This just gives it a little more context. When I first saw this, I wasn't sure what "recheck" meant at all.

File: src/libs/Authentication.js
Comment (tgolen) [inline src/libs/Authentication.js:60]: NAB: This seems like something we should be preventing from the front-end. Do you know what the actual length of "too few characters" is?

File: src/libs/Authentication.js
Comment (tgolen) [inline src/libs/Authentication.js:88]:
```suggestion
 * @param {String} [command] command name for logging purposes
```

File: src/libs/Middleware/Recheck.js
Comment (tgolen) [inline src/libs/Middleware/Recheck.js:26]:
```suggestion
        .finally(cancelRequestTimeoutTimer);
```

File: src/libs/Request.js
Comment (tgolen) [inline src/libs/Request.js:15]:
```suggestion
function makeXHR(request) {
```

File: src/libs/Request.js
Comment (tgolen) [inline src/libs/Request.js:25]:
```suggestion
function processWithMiddleware(request, isFromSequentialQueue = false) {
```

File: src/libs/API.js
Comment (tgolen) [inline src/libs/API.js:10]: It might be helpful to add some comments about the order they are used in and why that order matters.

**Review by tgolen — CHANGES_REQUESTED** [review]
I love the unit tests! Just some suggestions on names that might help make their purpose more clear.

File: src/libs/Authentication.js
Comment (marcaaron) [reply src/libs/Authentication.js:60]: Would have to look up the exact number. I believe we prevent it on the front end as well (would have to check on that as well though).

Comment (marcaaron) [reply src/libs/Authentication.js:60]: Looks like we are not https://github.com/Expensify/App/blob/84d058f5ade145a8bc89e00c36b340ba9078a8b0/src/pages/signin/PasswordForm.js#L84-L106 

Can create an issue for this.

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js]: yep, great suggestion

File: src/libs/API.js
Comment (marcaaron) [reply src/libs/API.js:10]: Added some more context here.

File: src/libs/Authentication.js
Comment (marcaaron) [reply src/libs/Authentication.js:60]: https://github.com/Expensify/App/issues/8794

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:27]: Technically, would not create a dependency cycle, but gonna just leave this in NetworkStore for now.

**Review by tylerkaraszewski — COMMENTED** [review]
I won't pretend to be an expert in this code, but I did read all of it. Only had a couple comments.

File: src/libs/Authentication.js
Comment (marcaaron) [reply src/libs/Authentication.js:141]: Good point and can see how another log could make this one ambiguous because we've already redirected.

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:36]: Yeah, this is a really great find and I think points to a deeper problem. 

The comment suggests that "Credentials haven't been initialized". So we are putting this request back into the queue because we are "waiting for the credentials". Which begs these questions: 

- Why we are waiting for them at all ?
- Why we are not waiting for them in the sequential queue ?

Gonna think about this for a second. My gut says no requests should happen at all until we can guarantee the credentials either exist or don't exist. Then we can just throw here and the user should probably also get signed out.

Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:36]: Looked into this a bit. The changes required here are a bit involved but I think worth doing...

We have some logic here:

https://github.com/Expensify/App/blob/58a139ab96a6189a6e0dcbbd003af18e7d5194dc/src/libs/Network/MainQueue.js#L23-L27

and it needs a better design since both the "main queue" and "sequential queue" should only make requests when we are "ready" to do so.

File: src/libs/Authentication.js
Comment (thienlnam) [inline src/libs/Authentication.js:126]: What happens if we fail authentication and unpause the network?  Does everything else just fail after with invalid authentication, should we also just cancel the rest of the requests here?

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:36]: Going to create a follow up issue for this as it's slightly tricky to untangle and I don't think it needs to be a blocker.

Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:36]: TL;DR - no, credentials are not automatically regenerated and this code appears related to a race condition. the queue we have for these requests now isn't "waiting for the credentials" so it's not clear if this is a problem or not, but should be cleaned up.

**Review by thienlnam — DISMISSED** [review]
Also not familiar with network code but the changes look great! Just a couple questions

File: src/libs/Middleware/Reauthentication.js
Comment (marcaaron) [reply src/libs/Middleware/Reauthentication.js:36]: Created an issue here: https://github.com/Expensify/Expensify/issues/208303

File: src/libs/Authentication.js
Comment (marcaaron) [reply src/libs/Authentication.js:101]: yeah, AFAIK we are not yet using any keychain features. Maybe using IndexedDB is slightly discouraging hackers here? Gonna resolve this though since it's unrelated to these changes.

File: src/libs/Authentication.js
Comment (marcaaron) [reply src/libs/Authentication.js:126]: It depends on why it failed and which queue is processing so we have to look at the `Retry` middleware to figure it out.

If we are in the main queue and this failed because of a networking issue then requests will get put back into the queue and will either run again when we come back online (if we went offline) or if we are online they will run again until we run out of retries. **Note:** That behavior will change soon because we will soon stop retrying basically "all requests" like we do now.

https://github.com/Expensify/App/blob/f1c8b4fa4570cee4820c079d8276243e8a350659/src/libs/Middleware/Retry.js#L58-L68

If we are in the sequential queue then we will leave the request in the queue but move onto the next one.

https://github.com/Expensify/App/blob/f72c7d510e522587dffcdc72464c153d1d7d69e2/src/libs/Network/SequentialQueue.js#L28-L33

On the other hand if this fails because we couldn't authenticate at all (and there are no networking issues) then the user must be logged out.

And the act of logging out will cancel the requests in the main queue and clear storage (which I _think_ would also take care of any persisted requests that haven't run yet).

https://github.com/Expensify/App/blob/f72c7d510e522587dffcdc72464c153d1d7d69e2/src/libs/actions/SignInRedirect.js#L50-L52

Comment (thienlnam) [reply src/libs/Authentication.js:126]: Ah gotcha, sweet thanks!

**Review by thienlnam — APPROVED** [review]

**Review by tgolen — APPROVED** [review]

**Comment by marcaaron** [comment]
Gonna merge since we have a couple of approvals. 

@tylerkaraszewski let me know if you have any thoughts about my response [here](https://github.com/Expensify/App/pull/8742#discussion_r860096488).

File: src/libs/Middleware/Reauthentication.js
Comment (tylerkaraszewski) [reply src/libs/Middleware/Reauthentication.js:36]: 👍

File: src/libs/Middleware/Reauthentication.js
Comment (Santhosh-Sellavel) [inline src/libs/Middleware/Reauthentication.js:24]: This PR causes a regression [here](https://github.com/Expensify/App/issues/16637) more details on the [PR](https://github.com/Expensify/App/pull/18237/files)

**Pull request merged**
